File: kerb_chain/playbook.py
"""
Playbook DSL — small, deterministic, auditable.

A playbook is a YAML file with a ``name`` and a list of ``plays``. Each
play has:

  name:        Free-form identifier shown in the runner's output.
  description: Human note for operators reviewing the playbook.
  when:        A simple condition expression (see ``conditions.py``)
               — the play runs once per *finding* that matches.
  command:     Either a list of argv, or a shell string. Placeholders
               (``{{domain}}``, ``{{finding.target}}``, etc.) are
               substituted from the Engagement context just before exec.
  capture:     Rules for parsing stdout/stderr/files into Loot.
  on_success:  List of follow-up play names to enqueue.
  requires_aggressive: Bool gate — operator passes ``--aggressive`` to
               kerb-chain to opt in to noisy plays (network spray,
               authentication coercion, anything that creates AD objects).
  category:    Free-form tag, used by ``--only-category`` filters.

The DSL deliberately has *no* loops, no variable assignment, no
sub-templating beyond placeholder substitution. Anything more complex
should be a real Python module wired in alongside.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml


@dataclass
class CaptureRule:
    """How a play teaches the runner to extract loot from its output.

    All fields are optional — a play with no capture rules just runs
    for side effects.
    """

    # Save stdout to a file in run_dir. Useful as a checkpoint or as
    # input to a follow-up play.
    stdout_to_file: str | None = None

    # Regex applied per-line over stdout; each match produces a
    # Credential whose username/password come from the named groups.
    cred_regex:        str | None = None    # must define groups 'user' and 'pass'
    cred_hash_regex:   str | None = None    # must define groups 'user' and 'hash'

    # Glob pattern under run_dir; every match becomes a Path entry in
    # loot.files keyed by the basename.
    files_glob: str | None = None

    # Mark a host as owned when stdout contains this substring.
    owned_marker: str | None = None
    owned_host:   str | None = None         # the host to record (template-rendered)


@dataclass
class Play:
    name:                str
    command:             list[str] | str    # argv list (preferred) or shell string
    description:         str = ""
    when:                str = ""           # empty = always
    per:                 str = "engagement"  # 'engagement' or 'finding'
    capture:             CaptureRule = field(default_factory=CaptureRule)
    on_success:          list[str] = field(default_factory=list)
    requires_aggressive: bool = False
    category:            str = "enumeration"
    timeout:             int = 600          # seconds


@dataclass
class Playbook:
    name:        str
    plays:       list[Play]
    description: str = ""
    path:        Path | None = None

    def by_name(self, name: str) -> Play | None:
        for p in self.plays:
            if p.name == name:
                return p
        return None

    @classmethod
    def from_file(cls, path: str | Path) -> Playbook:
        p = Path(path)
        data = yaml.safe_load(p.read_text())
        if not isinstance(data, dict):
            raise ValueError(f"{p}: top-level must be a mapping")

        raw_plays = data.get("plays") or []
        plays: list[Play] = []
        for i, item in enumerate(raw_plays):
            if "name" not in item:
                raise ValueError(f"{p}: play #{i} missing 'name'")
            if "command" not in item:
                raise ValueError(f"{p}: play '{item['name']}' missing 'command'")
            cap = CaptureRule(**(item.get("capture") or {}))
            plays.append(Play(
                name=item["name"],
                description=item.get("description", ""),
                command=item["command"],
                when=item.get("when", ""),
                per=item.get("per", "engagement"),
                capture=cap,
                on_success=list(item.get("on_success", []) or []),
                requires_aggressive=bool(item.get("requires_aggressive", False)),
                category=item.get("category", "enumeration"),
                timeout=int(item.get("timeout", 600)),
            ))
        return cls(
            name=data.get("name", p.stem),
            description=data.get("description", ""),
            plays=plays,
            path=p,
        )


# ────────────────────────────────────────────────────────────────────── #
#  Condition language                                                    #
# ────────────────────────────────────────────────────────────────────── #


def evaluate_condition(expr: str, *, finding: dict | None, engagement: Any) -> bool:
    """A tiny, hand-rolled condition language. No eval(), no exec().

    Supported forms (all return bool, all are case-insensitive on the LHS
    where appropriate):

      ``finding.attack == 'Kerberoast'``
      ``finding.attack in ['DCSync (full)', 'DCSync (partial)']``
      ``finding.severity in ['CRITICAL', 'HIGH']``
      ``finding.data.encryption == 'RC4'``
      ``loot.has_credential``
      ``loot.has_credential_for finding.target``
      ``not loot.has_credential``

    Multiple clauses joined by ``and`` / ``or`` (left-associative, no
    parens). Empty expression evaluates to True.
    """
    if not expr or not expr.strip():
        return True

    # Split on `and`/`or` while remembering the operator order.
    tokens = _tokenise_logical(expr)
    if not tokens:
        return True

    # Evaluate the first clause, then fold subsequent ones with the
    # operator that preceded them.
    result = _eval_clause(tokens[0], finding=finding, engagement=engagement)
    i = 1
    while i + 1 < len(tokens):
        op = tokens[i].lower()
        clause = tokens[i + 1]
        rhs = _eval_clause(clause, finding=finding, engagement=engagement)
        result = (result and rhs) if op == "and" else (result or rhs)
        i += 2
    return result


def _tokenise_logical(expr: str) -> list[str]:
    """Split a condition string on the keywords ``and`` / ``or`` while
    keeping the operators as separate tokens. Quoted substrings are
    preserved verbatim so something like ``foo == 'a and b'`` doesn't
    get split on the inner ``and``."""
    out: list[str] = []
    buf: list[str] = []
    i, n = 0, len(expr)
    in_quote: str | None = None
    while i < n:
        c = expr[i]
        if in_quote:
            buf.append(c)
            if c == in_quote:
                in_quote = None
            i += 1
            continue
        if c in "'\"":
            in_quote = c
            buf.append(c)
            i += 1
            continue
        # Try to match `and `/`or ` boundaries with whitespace.
        rest = expr[i:].lstrip()
        if rest.lower().startswith("and ") and (not buf or buf[-1].isspace()):
            out.append("".join(buf).strip())
            out.append("and")
            buf = []
            i = expr.index(rest, i) + 3
            continue
        if rest.lower().startswith("or ") and (not buf or buf[-1].isspace()):
            out.append("".join(buf).strip())
            out.append("or")
            buf = []
            i = expr.index(rest, i) + 2
            continue
        buf.append(c)
        i += 1
    last = "".join(buf).strip()
    if last:
        out.append(last)
    return [t for t in out if t]


def _eval_clause(clause: str, *, finding: dict | None, engagement: Any) -> bool:
    c = clause.strip()
    if c.startswith("not "):
        return not _eval_clause(c[4:], finding=finding, engagement=engagement)

    # loot predicates
    if c == "loot.has_credential":
        return bool(engagement.loot.credentials)
    if c.startswith("loot.has_credential_for "):
        target_expr = c[len("loot.has_credential_for "):].strip()
        target = _resolve_value(target_expr, finding=finding, engagement=engagement)
        if not isinstance(target, str):
            return False
        return engagement.loot.has_creds_for(target)

    # comparisons: lhs == 'literal' | lhs in [...]
    if " == " in c:
        lhs, rhs = c.split(" == ", 1)
        return _resolve_value(lhs, finding=finding, engagement=engagement) == _parse_literal(rhs)
    if " != " in c:
        lhs, rhs = c.split(" != ", 1)
        return _resolve_value(lhs, finding=finding, engagement=engagement) != _parse_literal(rhs)
    if " in " in c:
        lhs, rhs = c.split(" in ", 1)
        rhs_val = _parse_literal(rhs)
        if not isinstance(rhs_val, (list, tuple, set)):
            return False
        return _resolve_value(lhs, finding=finding, engagement=engagement) in rhs_val

    # bare identifier — truthy resolution
    return bool(_resolve_value(c, finding=finding, engagement=engagement))


def _resolve_value(token: str, *, finding: dict | None, engagement: Any) -> Any:
    token = token.strip()
    if token.startswith("finding."):
        if finding is None:
            return None
        path = token.split(".")[1:]
        node: Any = finding
        for part in path:
            if isinstance(node, dict):
                node = node.get(part)
            else:
                return None
        return node
    if token.startswith("loot.credentials"):
        return engagement.loot.credentials
    return token  # treat as literal-ish


def _parse_literal(token: str) -> Any:
    t = token.strip()
    if (t.startswith("'") and t.endswith("'")) or (t.startswith('"') and t.endswith('"')):
        return t[1:-1]
    if t.startswith("[") and t.endswith("]"):
        # Naïve list parser — splits on commas, respects single quotes.
        body = t[1:-1].strip()
        if not body:
            return []
        parts: list[str] = []
        buf: list[str] = []
        in_q: str | None = None
        for ch in body:
            if in_q:
                buf.append(ch)
                if ch == in_q:
                    in_q = None
                continue
            if ch in "'\"":
                in_q = ch
                buf.append(ch)
                continue
            if ch == ",":
                parts.append("".join(buf).strip())
                buf = []
                continue
            buf.append(ch)
        if buf:
            parts.append("".join(buf).strip())
        return [_parse_literal(p) for p in parts]
    # numeric
    try:
        if "." in t:
            return float(t)
        return int(t)
    except ValueError:
        return t  # bare identifier

File: kerb-chain-rs/src/playbook.rs
//! Playbook DSL — small, deterministic, no eval.
//!
//! Direct port of `kerb_chain/playbook.py`. Same YAML shape, same
//! condition grammar, same defaults. The Python and Rust runtimes can
//! load each other's playbooks without translation.

use std::path::{Path, PathBuf};

use anyhow::{Context, Result, anyhow};
use serde::{Deserialize, Serialize};

use crate::engagement::Engagement;
use crate::findings::Finding;


// ─────────────────────────────────────────────────────────────────── //
//  Schema                                                             //
// ─────────────────────────────────────────────────────────────────── //


#[derive(Debug, Clone, Default, Deserialize, Serialize)]
pub struct CaptureRule {
    #[serde(default)]
    pub stdout_to_file:  Option<String>,
    #[serde(default)]
    pub cred_regex:      Option<String>,
    #[serde(default)]
    pub cred_hash_regex: Option<String>,
    #[serde(default)]
    pub files_glob:      Option<String>,
    #[serde(default)]
    pub owned_marker:    Option<String>,
    #[serde(default)]
    pub owned_host:      Option<String>,
}


/// A play's `command` field is either an argv list (preferred — no
/// shell interpretation) or a single shell string (rendered then
/// shell-split with shell-words to preserve quoting).
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(untagged)]
pub enum Command {
    Argv(Vec<String>),
    Shell(String),
}


fn default_per()      -> String { "engagement".to_string() }
fn default_category() -> String { "enumeration".to_string() }
fn default_timeout()  -> u64    { 600 }


#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Play {
    pub name:        String,
    #[serde(default)]
    pub description: String,
    pub command:     Command,
    #[serde(default)]
    pub when:        String,
    #[serde(default = "default_per")]
    pub per:         String, // "engagement" or "finding"
    #[serde(default)]
    pub capture:     CaptureRule,
    #[serde(default)]
    pub on_success:  Vec<String>,
    #[serde(default)]
    pub requires_aggressive: bool,
    #[serde(default = "default_category")]
    pub category:    String,
    #[serde(default = "default_timeout")]
    pub timeout:     u64,
}


#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Playbook {
    #[serde(default)]
    pub name:        String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub plays:       Vec<Play>,
    #[serde(skip)]
    pub path:        Option<PathBuf>,
}

impl Playbook {
    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
        let p = path.as_ref();
        let text = std::fs::read_to_string(p)
            .with_context(|| format!("reading {}", p.display()))?;
        let mut pb: Playbook = serde_yaml::from_str(&text)
            .with_context(|| format!("parsing {} as YAML", p.display()))?;

        // Validate plays — fail loudly on obvious typos rather than
        // silently dropping plays at runtime.
        for (i, play) in pb.plays.iter().enumerate() {
            if play.name.is_empty() {
                return Err(anyhow!("{}: play #{} missing 'name'", p.display(), i));
            }
            match &play.command {
                Command::Argv(v)  if v.is_empty() => {
                    return Err(anyhow!(
                        "{}: play '{}' has empty argv", p.display(), play.name));
                }
                Command::Shell(s) if s.trim().is_empty() => {
                    return Err(anyhow!(
                        "{}: play '{}' has empty shell command",
                        p.display(), play.name));
                }
                _ => {}
            }
        }

        if pb.name.is_empty() {
            pb.name = p.file_stem()
                .map(|s| s.to_string_lossy().into_owned())
                .unwrap_or_else(|| "playbook".to_string());
        }
        pb.path = Some(p.to_path_buf());
        Ok(pb)
    }

    pub fn by_name(&self, name: &str) -> Option<&Play> {
        self.plays.iter().find(|p| p.name == name)
    }
}


// ─────────────────────────────────────────────────────────────────── //
//  Condition language                                                 //
// ─────────────────────────────────────────────────────────────────── //


/// Evaluate a `when` clause. Returns `true` when:
///
/// - the expression is empty
/// - all `and`-joined sub-clauses match
/// - any `or`-joined sub-clause matches (left-associative, no parens)
///
/// Supported atoms (no eval, no exec):
///
/// ```text
/// finding.attack == 'Kerberoast'
/// finding.attack in ['DCSync (full)', 'DCSync (partial)']
/// finding.severity in ['CRITICAL', 'HIGH']
/// finding.data.encryption == 'RC4'
/// loot.has_credential
/// loot.has_credential_for finding.target
/// not loot.has_credential
/// ```
pub fn evaluate_condition(
    expr:       &str,
    finding:    Option<&Finding>,
    engagement: &Engagement,
) -> bool {
    if expr.trim().is_empty() {
        return true;
    }
    let tokens = tokenise_logical(expr);
    if tokens.is_empty() {
        return true;
    }

    let mut result = eval_clause(&tokens[0], finding, engagement);
    let mut i = 1;
    while i + 1 < tokens.len() {
        let op  = tokens[i].to_ascii_lowercase();
        let rhs = eval_clause(&tokens[i + 1], finding, engagement);
        result = match op.as_str() {
            "and" => result && rhs,
            "or"  => result || rhs,
            _     => result,  // unknown joiner — keep going, ignore
        };
        i += 2;
    }
    result
}


fn tokenise_logical(expr: &str) -> Vec<String> {
    let chars: Vec<char> = expr.chars().collect();
    let n = chars.len();
    let mut out: Vec<String> = Vec::new();
    let mut buf: String = String::new();
    let mut i = 0usize;
    let mut in_quote: Option<char> = None;

    while i < n {
        let c = chars[i];

        if let Some(q) = in_quote {
            buf.push(c);
            if c == q { in_quote = None; }
            i += 1;
            continue;
        }
        if c == '\'' || c == '"' {
            in_quote = Some(c);
            buf.push(c);
            i += 1;
            continue;
        }

        // Boundary check: only split on ` and ` / ` or ` (whitespace
        // before the keyword) so we don't break identifiers like
        // `command`. Equivalent to the Python tokeniser's `buf[-1].isspace()`
        // guard.
        let prev_is_ws = buf.chars().last().map(|c| c.is_whitespace()).unwrap_or(true);
        if prev_is_ws {
            let rest_lower: String = chars[i..].iter().collect::<String>().to_ascii_lowercase();
            if rest_lower.starts_with("and ") {
                out.push(buf.trim().to_string());
                out.push("and".to_string());
                buf.clear();
                i += 4;     // "and "
                continue;
            }
            if rest_lower.starts_with("or ") {
                out.push(buf.trim().to_string());
                out.push("or".to_string());
                buf.clear();
                i += 3;     // "or "
                continue;
            }
        }

        buf.push(c);
        i += 1;
    }

    let last = buf.trim().to_string();
    if !last.is_empty() {
        out.push(last);
    }
    out.into_iter().filter(|s| !s.is_empty()).collect()
}


fn eval_clause(clause: &str, finding: Option<&Finding>, engagement: &Engagement) -> bool {
    let c = clause.trim();
    if let Some(rest) = c.strip_prefix("not ") {
        return !eval_clause(rest, finding, engagement);
    }

    if c == "loot.has_credential" {
        return !engagement.loot.credentials.is_empty();
    }
    if let Some(rest) = c.strip_prefix("loot.has_credential_for ") {
        let target = resolve_value(rest.trim(), finding, engagement);
        return target
            .as_str()
            .map(|s| engagement.loot.has_creds_for(s))
            .unwrap_or(false);
    }

    if let Some((lhs, rhs)) = split_once_outside_quotes(c, " == ") {
        return resolve_value(lhs, finding, engagement) == parse_literal(rhs);
    }
    if let Some((lhs, rhs)) = split_once_outside_quotes(c, " != ") {
        return resolve_value(lhs, finding, engagement) != parse_literal(rhs);
    }
    if let Some((lhs, rhs)) = split_once_outside_quotes(c, " in ") {
        let lhs_v = resolve_value(lhs, finding, engagement);
        let rhs_v = parse_literal(rhs);
        if let Some(arr) = rhs_v.as_array() {
            return arr.iter().any(|v| v == &lhs_v);
        }
        return false;
    }

    // Bare identifier — truthy fallback
    matches!(resolve_value(c, finding, engagement),
             serde_json::Value::Bool(true))
}


/// Split on `sep` only when `sep` doesn't fall inside a quoted region.
/// Used so that `finding.target == 'a == b'` doesn't mis-split.
fn split_once_outside_quotes<'a>(s: &'a str, sep: &str) -> Option<(&'a str, &'a str)> {
    let bytes = s.as_bytes();
    let sep_bytes = sep.as_bytes();
    let mut in_quote: Option<u8> = None;
    let mut i = 0usize;
    while i + sep_bytes.len() <= bytes.len() {
        let b = bytes[i];
        if let Some(q) = in_quote {
            if b == q { in_quote = None; }
            i += 1;
            continue;
        }
        if b == b'\'' || b == b'"' {
            in_quote = Some(b);
            i += 1;
            continue;
        }
        if &bytes[i..i + sep_bytes.len()] == sep_bytes {
            return Some((&s[..i], &s[i + sep_bytes.len()..]));
        }
        i += 1;
    }
    None
}


fn resolve_value(
    token:      &str,
    finding:    Option<&Finding>,
    engagement: &Engagement,
) -> serde_json::Value {
    let t = token.trim();

    if let Some(rest) = t.strip_prefix("finding.") {
        let Some(f) = finding else { return serde_json::Value::Null };
        let mut node: serde_json::Value =
            serde_json::Value::Object(serde_json::Map::from_iter(f.clone()));
        for part in rest.split('.') {
            node = match node {
                serde_json::Value::Object(map) => {
                    map.get(part).cloned().unwrap_or(serde_json::Value::Null)
                }
                _ => serde_json::Value::Null,
            };
        }
        return node;
    }

    if t == "loot.credentials" {
        return serde_json::Value::Bool(!engagement.loot.credentials.is_empty());
    }

    serde_json::Value::String(t.to_string())
}


fn parse_literal(token: &str) -> serde_json::Value {
    let t = token.trim();
    if (t.starts_with('\'') && t.ends_with('\'')) || (t.starts_with('"') && t.ends_with('"')) {
        return serde_json::Value::String(t[1..t.len() - 1].to_string());
    }
    if t.starts_with('[') && t.ends_with(']') {
        let body = t[1..t.len() - 1].trim();
        if body.is_empty() {
            return serde_json::Value::Array(Vec::new());
        }
        let mut parts: Vec<String> = Vec::new();
        let mut buf = String::new();
        let mut in_q: Option<char> = None;
        for ch in body.chars() {
            if let Some(q) = in_q {
                buf.push(ch);
                if ch == q { in_q = None; }
                continue;
            }
            if ch == '\'' || ch == '"' {
                in_q = Some(ch);
                buf.push(ch);
                continue;
            }
            if ch == ',' {
                parts.push(buf.trim().to_string());
                buf.clear();
                continue;
            }
            buf.push(ch);
        }
        if !buf.is_empty() { parts.push(buf.trim().to_string()); }
        return serde_json::Value::Array(parts.into_iter().map(|p| parse_literal(&p)).collect());
    }
    if let Ok(n) = t.parse::<i64>() { return serde_json::Value::from(n); }
    if let Ok(f) = t.parse::<f64>() { return serde_json::Value::from(f); }

    // Bare identifier — treat as a string.
    serde_json::Value::String(t.to_string())
}
